src/treetensornetworks/projttns/projttn_apply.jl

```diff
@@ -1,57 +1,114 @@
 struct ProjTTNApply{V} <: AbstractProjTTN{V}
   pos::Union{Vector{<:V},NamedEdge{V}}
-  psi0::TTN{V}
-  H::TTN{V}
+  init_state::AbstractTTN{V}
+  operator::AbstractTTN{V}
   environments::Dictionary{NamedEdge{V},ITensor}
 end
 
-function ProjTTNApply(psi0::TTN, H::TTN)
-  return ProjTTNApply(vertextype(H)[], psi0, H, Dictionary{edgetype(H),ITensor}())
+init_state(p::ProjTTNApply) = p.init_state
+
+function ProjTTNApply(init_state::AbstractTTN, operator::AbstractTTN)
+  return ProjTTNApply(
+    vertextype(operator)[], init_state, operator, Dictionary{edgetype(operator),ITensor}()
+  )
 end
 
 function copy(P::ProjTTNApply)
-  return ProjTTNApply(P.pos, copy(P.psi0), copy(P.H), copy(P.environments))
+  return ProjTTNApply(
+    pos(P), copy(init_state(P)), copy(operator(P)), copy_keys_values(environments(P))
+  )
 end
 
+function unsafe_copy(P::ProjTTNApply)
+  return ProjTTNApply(P.pos, copy(init_state(P)), copy(operator(P)), copy(environments(P)))
+end
 function set_nsite(P::ProjTTNApply, nsite)
   return P
 end
 
 function shift_position(P::ProjTTNApply, pos)
-  return ProjTTNApply(pos, P.psi0, P.H, P.environments)
+  return ProjTTNApply(pos, init_state(P), operator(P), environments(P))
+end
+
+function set_environments(p::ProjTTNApply, environments)
+  return ProjTTNApply(pos(p), init_state(p), operator(p), environments)
+end
+
+function set_environment(p::ProjTTNApply, edge, env)
+  newenv = merge(p.environments, Dictionary((edge,), (env,)))
```

Comment:
```suggestion
  newenv = merge(environments(p), Dictionary((edge,), (env,)))
```

Reply:
Replaced this by call to inplace with copy.